crates/type-syntax/src/parser/internal/object.rs

```diff
@@ -0,0 +1,230 @@
+        }
+    };
+
+    Ok(Type::Shape(ShapeType {
+        kind,
+        keyword,
+        left_brace: stream.eat(TypeTokenKind::LeftBrace)?.span,
+        fields: {
+            let mut fields = Vec::new();
+            while !stream.is_at(TypeTokenKind::RightBrace)? && !stream.is_at(TypeTokenKind::Ellipsis)? {
+                let has_key = {
+                    let mut found_key = false;
+                    // Scan ahead to determine if a key is present before the value type.
+                    for i in 0.. {
+                        let Some(token) = stream.lookahead(i)? else {
+                            // Reached the end of the stream, so no key was found.
+                            break;
+                        };
+
+                        match token.kind {
+                            // If we find a colon, we know a key is present.
+                            TypeTokenKind::Colon => {
+                                found_key = true;
+                                break;
+                            }
+                            TypeTokenKind::Question => {
+                                // If we find a question mark, it could indicate a key,
+                                // if the following token is a colon.
+                                if stream.lookahead(i + 1)?.is_some_and(|t| t.kind == TypeTokenKind::Colon) {
+                                    found_key = true;
+                                    break;
+                                } else {
+                                    // If the question mark is not followed by a colon,
+                                    // it could be part of the key.
+                                    continue;
+                                }
+                            }
+                            // If we find any of these tokens, what came before must have
+                            // been a full value type, not a key.
+                            TypeTokenKind::Comma
+                            | TypeTokenKind::RightBrace
+                            | TypeTokenKind::LeftBrace
+                            | TypeTokenKind::LeftParenthesis
+                            | TypeTokenKind::RightParenthesis
+                            | TypeTokenKind::LeftBracket
+                            | TypeTokenKind::RightBracket
+                            | TypeTokenKind::Ellipsis => {
+                                break;
+                            }
+                            // Any other token is part of a potential key, so keep scanning.
+                            _ => continue,
+                        }
+                    }
+
+                    found_key
+                };
+
+                let field = ShapeField {
+                    key: if has_key {
+                        Some(ShapeFieldKey {
+                            key: parse_shape_field_key(stream)?,
```

Comment:
object keys can only be identifiers probably. or tokens that are keywords ( but don't contain `-` )